### ISSUE TITLE
system date is not same with windows (WSL 2)

### ISSUE THREAD

**Issue by erdemece** [microsoft/WSL#4245]
* Your Windows build number:  18922

* date is updating itself in WSL 2 even I set the timezone and change date manually with date command

* because the date is wrong I can't sudo apt update and I have to set the date to my local time everytime I open WSL 2

thanks.

**Comment by benhillis** [comment]
Which distro are you using?  Any errors in dmesg? WSL relies on the tzdata package to do the translation.

**Comment by erdemece** [comment]
> Which distro are you using? Any errors in dmesg? WSL relies on the tzdata package to do the translation.

it's ubuntu 18.04
well I just converted from WSL 2 to WSL 1. I will reply after I convert WSL 2 with the outputs.

Thanks.

**Comment by radu-matei** [comment]
I can also confirm this, and it’s especially bad when changing time zones, leading up to not being able to use a package manager at all.

**Comment by BlaQPhoeniX** [comment]
```
[    0.335041] rtc_cmos 00:00: setting system clock to 2019-06-21 22:54:24 UTC (1561157664)
[    0.335057] Unstable clock detected, switching default tracing clock to "global"
               If you want to keep using the local clock, then add:
                 "trace_clock=local"
               on the kernel command line
```

This is what i see in dmesg related to system time. I'm using the Ubuntu 18.04 WSL 2.

**Comment by lmatter** [comment]
Me too.

**Comment by cxsy** [comment]
```
E: Release file for http://security.ubuntu.com/ubuntu/dists/bionic-security/InRelease is not valid yet (invalid for another 14h 47min 54s). Updates for this repository will not be applied.
E: Release file for http://archive.ubuntu.com/ubuntu/dists/bionic-updates/InRelease is not valid yet (invalid for another 15h 2min 9s). Updates for this repository will not be applied.
E: Release file for http://archive.ubuntu.com/ubuntu/dists/bionic-backports/InRelease is not valid yet (invalid for another 14h 49min 15s). Updates for this repository will not be applied.
```
This is the error message I got when I run `sudo apt update`. I think it's related to the wrong system date.
One workaround is switch to WSL1 and then switch back.

**Comment by ngasst** [comment]
I was having the same issue, but my main gripe was that it was preventing me from reliably working with docker containers in what was otherwise a very workable setup.

Here's my solution along with necessary code snippet.

1) Get time from reliable source on the internet using Curl.
2) Parse time and set it onto the WSL linux instance
3) Create a shell script and modify  .rc file to run said script at start

Script:
```sh
# ~/set-date.sh

# Get time from google server or some other reliable source
TIME_FROM_SERVER=$(curl -v --insecure --silent https://google.com/ 2>&1 | grep Date | sed -e 's/< Date: //');

# Set stored time
date +"%d%m%Y%H%M%S" -d "$TIME_FROM_SERVER";

echo "Time updated with success!";
```
After you run this, you should see the current timestamp printed as well as the success message.

Then, in your `bashrc`, `zshrc` or equivalent, call the script
`sh ~/set-date.sh`

CREDIT: I was having trouble with the returned time format. The script above is a barely modified version from [here](https://askubuntu.com/questions/741298/how-to-get-datetime-using-curl-command)

**Comment by MiklerGM** [comment]
The problem with the date begins when I put my notebook at sleep, the system clock is ticking but not the WSL one. It can be easily reproduced.

I've been using `ntpdate` for time sync purposes.
```
sudo ntpdate pool.ntp.org
```

**Comment by WSLUser** [comment]
I would venture using [chrony](https://linux.die.net/man/1/chrony) would be a better option than ntp. Or even [ntpsec](https://www.ntpsec.org/). Just add something similar such as `pool 2.fedora.pool.ntp.org iburst`to `chrony.conf` (or `ntp.conf`). If you have a local NTP server serving up time for all your machines, point there (and please have 2 servers configured. Yes one will work, but it is very bad practice).

**Comment by peidaqi** [comment]
Can confirm. This is a very serious bug, which means any date/time related service would be unreliable in WSL2... including apt.

**Comment by Rhahkeem** [comment]
Can confirm this as well. On converted distro's and newly downloaded ones. This makes it impossible to download newer packages from `apt`

**Comment by jdub** [comment]
Quick fix! In PowerShell or Command Prompt:
```
wsl --shutdown
```

Next time you start a WSL session, it'll be a cold boot and the time will be correct.

**Comment by benhillis** [comment]
We are working on a fix, in the meantime "sudo hwclock -s" should get things working.

**Comment by nunix** [comment]
just a quick and concrete example of when it happened for me:
- started a WSL2 distro
- made the laptop sleep
- started again the laptop the next day
- checked the date/time -> was still set to the day before

I went the `wsl --shutdown` route and it worked. Will do the `hwclock` next time 👍

**Comment by pvnguyen123** [comment]
Glad i found this post.  This was annoying me for awhile, FYI this also affects azure command line due to bad timestamp.

This works for me:

> sudo hwclock -s

Below is the error message from az cli in case someone is searching.  The error is misleading, but the root cause is that the CLI is using an outdated sys time from WSL2.  
---
You do not have the required permissions needed to perform this operation.
Depending on your operation, you may need to be assigned one of the following roles:
    "Storage Blob Data Contributor (Preview)"
    "Storage Blob Data Reader (Preview)"
    "Storage Queue Data Contributor (Preview)"
    "Storage Queue Data Reader (Preview)"

**Comment by bertonha** [comment]
I had this exactly same problem back on the time that I was using VirtualBox for virtualization, every time my host machine go to sleep. The virtual machine stopped the clock on that time. But when the host come back from sleep, the virtual machine does not look at the time on host machine to update the clock. This simple continue counting the time from where it had stopped.

Exactly the same bug again.

**Comment by benhillis** [comment]
This should be much better in build [18970](https://github.com/MicrosoftDocs/WSL/releases/tag/18970).

**Comment by onomatopellan** [comment]
@benhillis I thought pausing/unpausing a VM would be the equivalent to sleep a PC but I guess I was wrong. Would it be possible a similar fix when pausing/unpausing a VM?

**Comment by Biswa96** [comment]
What's the relation between pause/resume of a VM with this issue which is fixed now?

**Comment by onomatopellan** [comment]
@Biswa96 It's the same problem. If you pause a VM for some time then after unpausing it system date is not the same and `apt update` shows [this error](https://github.com/microsoft/WSL/issues/4245#issuecomment-507517006) .

**Comment by Biswa96** [comment]
There is a way. I'll add the code in [hvtool](https://github.com/Biswa96/hvtool).

**Comment by Biswa96** [comment]
There is a way to pause/resume WSL2 VM. I'll add the code in [hvtool](https://github.com/Biswa96/hvtool).

**Comment by Biswa96** [comment]
There is a semi-documented way to pause/resume WSL2 VM. The code is in [hvtool](https://github.com/Biswa96/hvtool). I have tried WSL VM to pause and resume but the time is in sync.

**Comment by onomatopellan** [comment]
That's the container. I mean pausing/resuming (2) the virtual machine that contains build 18970 reproduces this issue.
![](https://www.tenforums.com/attachments/virtualization/101037d1485971752t-suspend-resume-vm-much-slower-hyper-v-than-vmware-image.png)

**Comment by benhillis** [comment]
@onomatopellan - Looks like this fix doesn't work with nested, investigating...

**Comment by Kai-Richardson** [comment]
This has been fixed on my machine using previous error cases (sleeping) as of build 18980.

**Comment by ricardosantos9521** [comment]
This is happening in build 19002 after the pc is suspended for long hours. wsl.exe --shutdown is the fix

**Comment by johnmcdowall** [comment]
Just ran into this in build 19008.1000. `sudo hwclock -s` fixed it.

**Comment by ricardosantos9521** [comment]
I think this is not fixed, since i have to do ```script  sudo hwclock -s```  almost everyday morning. So please fix this. This happen after my surface laptop 1 is suspended for multiple hours.

![image](https://user-images.githubusercontent.com/20097801/68466953-45f90700-020d-11ea-9aa3-e69b8d4ce175.png)

![image](https://user-images.githubusercontent.com/20097801/68466963-4a252480-020d-11ea-8f47-2bac19697fa6.png)

**Comment by Rhahkeem** [comment]
Same as above . This still happens for me with vscode too

**Comment by benhillis** [comment]
Any output in dmesg?

**Comment by ricardosantos9521** [comment]
[dmseg.log](https://github.com/microsoft/WSL/files/3825542/dmseg.log)

**Comment by benhillis** [comment]
@ricardosantos9521 - thanks, those events are not relevant, anything about not being able to translate timezone?  What is the output of `readlink /etc/localtime`?

**Comment by ricardosantos9521** [comment]
No it doesn't appear nothing related to timezone translation.

readlink /etc/localtime returns (that is my timezone):
`/usr/share/zoneinfo/Europe/Lisbon`

**Comment by benhillis** [comment]
@ricardosantos9521 - Ok, and does `sudo hwclock -s` get you back in sync?

**Comment by ricardosantos9521** [comment]
Yes it does

**Comment by miklevin** [comment]
> Yes it does, until the next day morning

This is my experience too. I have to sudo hwclock -s every day. Not cool.

**Comment by benhillis** [comment]
@miklevin - In your case, how much does your clock drift?  To be clear, we're working on a fix for this, I'm just presenting some workarounds in the meantime (wsl.exe --shutdown is very heavy-handed).

**Comment by ricardosantos9521** [comment]
I want to add some new things. I think this is related to the way that suspension works (in this case on my surface laptop). After some hours of suspension (appears to be arround 4 hours) the pc goes to a kind of hibernation instead of continuing in suspension. So when i awake the pc in this 4 hours interval the clock appears to be ok. After that interval the clock starts to drift. 
So the drift in my case appears to be related to this 4 hours. If i suspended my pc at 1pm it appears that the clock will work until 5pm after that the clock starts to drift.

**Comment by benhillis** [comment]
@ricardosantos9521 - Thanks for the additional info.  If you don't mind, could you describe your config? Are you using WSL2 nested?  What kind of hardware are you using?

**Comment by ricardosantos9521** [comment]
I am runing Windows Fast Ring Builds as my main and unique os in a Surface Laptop 1 (i5-7200U) with all the default configs for WSL2.  I am running Ubuntu from the Store.

**Comment by johnmcdowall** [comment]
FWIW it was super clear to me that the amount of time that the Ubuntu clock drifted was exactly the amount of time my laptop had been closed shut and sleeping. On wake WSL is not re-syncing the clock for whatever reason. Running `sudo hwclock -s` fixes the time to be the correct time and is fine until the next time you sleep your machine.

**Comment by nicolashopin** [comment]
Same issue on a Dell XPS on the slow build. Definitely related to sleep

**Comment by timbophillips** [comment]
I have the same issue DEll XPS 13 7390 Windows Build 19030.1

**Comment by benhillis** [comment]
If somebody is able to reproduce this could you please run `powercfg /spr` and share the contents?

**Comment by nicolashopin** [comment]
@benhillis Here's the sleep study from just after my pc woke up.
Run `date` on ubuntu and was about ~ 1 day behind (consistent with the sleep time)
https://gist.github.com/nicolashopin/b472afb4d0667ee75e43ee134a3aac73

**Comment by marrobi** [comment]
Same here, 19033 - @benhillis have emailed you my report.

**Comment by noelbundick** [comment]
+1, another report in your inbox @benhillis

**Comment by nicolashopin** [comment]
It seems to be fixed now for me:
`19035.1` I recently got both a firmware and OS updates

**Comment by nicolashopin** [comment]
Dismiss the above

**Comment by miklevin** [comment]
Another report in your inbox @benhillis

**Comment by myxoh** [comment]
Any ETA on a potential fix?

**Comment by benhillis** [comment]
Issue is being debugged, in the meantime I suggest using "sudo hwclock -s" as needed.

**Comment by chrisdpratt** [comment]
Same issue here, but running `hwclock -s` gives an error: `hwclock: Cannot access the Hardware Clock via any known method.`. I can probably fix that by adding/updating the clocksource, but considering that this is happening in development containers, which should be able to be thrown away whenever, it's not a viable solution to have to run a series of commands to enable `hwclock` every time they're created. (For reference these are being built against the `mcr.microsoft.com/dotnet/core/aspnet:3.1-buster-slim` image.)

Anyways, as another suggested, running `wsl --shutdown` (and restarting Docker afterwards) resets the clock. So that's at least a good workaround until this is fixed.

**Comment by rqueraud** [comment]
I also have the error `hwclock: Cannot access the Hardware Clock via any known method.` trying to build a docker image within wsl and can confirm that restarting docker make it work again.

**Comment by jwhipp** [comment]
Ran into this today, used the ntpdate option. 

I wouldn't have noticed but the AWS cli was giving weird time discrepancy errors.

**Comment by sebader** [comment]
same here on Ubuntu 18.04, Windows 10 19041.153

**Comment by JaredReisinger** [comment]
@chrisdpratt and @rqueraud: I found that the `sudo` is important.  You'll get the "Cannot access the Hardware Clock" message with `hwclock -s`, but `sudo hwclock -s` works just fine.

**Comment by rossjrw** [comment]
It's no solution, but one trick I've implemented to make my life easier is adding this to my `~/.profile`:
```shell
PS1="\t $PS1"
```
which adds the current time to my command prompt. Then it's on me to remember to pay attention, compare the time to my system clock, and run `sudo hwclock -s` if appropriate - which all too often I still forget, and my commit history pays the price!

**Comment by chrisdpratt** [comment]
@JaredReisinger: I'm running in containers, so I'm root. The sudo command is only to escalate privilege. That's not necessary for the root account.

**Comment by tlfmcooper** [comment]
> Quick fix! In PowerShell or Command Prompt:
> 
> ```
> wsl --shutdown
> ```
> 
> Next time you start a WSL session, it'll be a cold boot and the time will be correct.

That worked for me, thanks

**Comment by speller** [comment]
Facing the same issue. When it will be fixed?

**Comment by onomatopellan** [comment]
@speller Doesn't `sudo hwclock -s` work for you?

**Comment by halink0803** [comment]
> @speller Doesn't `sudo hwclock -s` work for you?

I am facing this issue. `sudo hwclock -s` return this: 

```
hwclock: Cannot access the Hardware Clock via any known method.
hwclock: Use the --debug option to see the details of our search for an access method.
```

How can I update time for my wsl?

**Comment by onomatopellan** [comment]
@halink0803 Can you paste what you see after entering these:
```bash
sudo hwclock -s --debug
ll /dev/rtc*
sudo dmesg | grep rtc
```

**Comment by speller** [comment]
@onomatopellan 
> Doesn't sudo hwclock -s work for you?

It works. No errors, just updated the current date:

![image](https://user-images.githubusercontent.com/1829515/79521530-5ce13680-8094-11ea-93bf-c5169b70b91f.png)

**Comment by hwine** [comment]
FWIW - I've done a script to make this easier. (I'm on a laptop, where it happens every time the machine sleeps).

Details in [this gist](https://gist.github.com/hwine/51f7d941cc3c59da35add92da1408604).

Hope it can help others.

**Comment by eteq** [comment]
I confirm seeing the same thing whenever my laptop sleeps.  Does anyone know if there's a way for WSL to get a message that the machine has gone to sleep?  Presumably then one could have that trigger trigger `hwclock`. Otherwise it seems like the only solution that doesn't require remembering to run a command (which I have consistently failed to do...) is to use `cron` or similar...

**Comment by onomatopellan** [comment]
@eteq This script runs whenever the PC comes from sleep/hibernate
https://github.com/stuartleeks/wsl-clock

**Comment by stuartleeks** [comment]
Thanks @onomatopellan !

I've had that script/task set up for a couple of days and it has been working well.

Would love to get confirmation if it's working for others, too 😃

And I've written a blog post with a bit more of a breakdown of the script: https://stuartleeks.com/posts/fixing-clock-skew-with-wsl-2/

**Comment by onomatopellan** [comment]
@stuartleeks Thank you for the script! I never had this problem in my PC though, but I can see it when pausing/unpausing my hyper-v VMs. Unfortunately the script won't help on VMs because hwclock doesn't work and doesn't seem to trigger any event when unpausing the VM.
For now inside a VM I use ntpdate to sync the clock with internet.

**Comment by einarpersson** [comment]
This suddenly happened to me today. WSL2 Ubuntu 18.04 still thought it was friday (I wish!) but it was monday 🤷‍♂️ 

`sudo hwclock -s` worked as a temporary fix, but the fact that this issue has been open for a year makes me wonder

Running Windows 10, v 2004, build 19041.207 on a Dell XPS 9380

**Comment by edburns** [comment]
The `sudo hwclock -s` doesn't work for me. Even as a temporary fix.

**Comment by edburns** [comment]
This also messes with OAuth 2.0 if you are running a demo webapp locally that happens to use that.

**Comment by garyo** [comment]
This is happening to me too. Just upgraded a WSL1 Ubuntu 18.04 to WSL2, left the laptop for a day, and it still thinks it's yesterday in Linux-land.

**Comment by garyo** [comment]
@therealkenc I am running the release build, does not seem to be fixed for me. Is this fixed in an upcoming version?

**Comment by therealkenc** [comment]
Something became much better in 18970 circa [August](https://github.com/microsoft/WSL/issues/4245#issuecomment-526363427) 2019, which means that fix shipped in 2004 aka 19041. That thing might not be your thing; feel free to revisit in another issue with your repro.

**Comment by garyo** [comment]
OK, but it's _exactly_ the same as this issue so not sure if opening a new is is best. After machine wakes up from sleep or hibernate, date is off and is not corrected until I use `sudo hwclock -s` or some other temporary fix (e.g. shut down & restart the VM). Does not happen with WSL 1. Seems from comments above like I'm not the only one still experiencing this.

**Comment by dwaynebradley** [comment]
@garyo You're not the only one still seeing this. I just ran into this same issue today myself. My Ubuntu 20.04 environment in WSL2 is way off after I put my laptop to sleep last night.

I wonder if there is a difference in

1. Installing a new Ubuntu 20.04 environment after upgrading to WSL2 vs
2. upgrading an Ubuntu environment after upgrading WSL1->WSL2

I did (2) above with my setup.

**Comment by padevel** [comment]
@dwaynebradley Same problem here with '2' (Upgrade of Debain WSL1->2).

sudo hwclock -s works, but is pretty tedious/haphazard. This laptop is alseep a handful of time per day.

**Comment by skoging** [comment]
> @garyo You're not the only one still seeing this. I just ran into this same issue today myself. My Ubuntu 20.04 environment in WSL2 is way off after I put my laptop to sleep last night.
> 
> I wonder if there is a difference in
> 
> 1. Installing a new Ubuntu 20.04 environment after upgrading to WSL2 vs
> 2. upgrading an Ubuntu environment after upgrading WSL1->WSL2
> 
> I did (2) above with my setup.
> 
> Running `sudo hwclock -s` updates my time but is a pain to remember to do until it is too late.

Happens with 1. as well. Arrived here in this thread when I ran into issues installing the dotnet SDK, after the laptop had been asleep for a couple days. `sudo hwclock -s` works to update the date and time.

**Comment by stuartleeks** [comment]
Just going to pop another mention of this here: https://github.com/stuartleeks/wsl-clock
Since setting that up, I've forgotten about the problem (apart from notifications from this issue thread) :-)

**Comment by dwaynebradley** [comment]
Well, even after I run `sudo hwclock -s` now, the time in Ubuntu 20.04 is still approximately 3 seconds behind what Windows is showing. If I want it close, I actually have to use `sudo date -s "YYYY-MM-DD HH24:MI:SS"`. Hopefully this is corrected soon.

**Comment by garyo** [comment]
You can run `ntp` (via systemd) which will sync the clock, as long as it's reasonably close to start. It may take a while to resync after the hiccup though, not sure. It only slews the clock pretty gradually.

**Comment by LaurentLesle** [comment]
running `sudo hwclock -s` from wsl2 when resuming from sleep mode resynch the clock in the containers. At least I don't have to reboot Windows 10.

**Comment by mileslane** [comment]
This is happening to me as well.  Ubuntu 20.04 in WSL2 (latest Win10 insider updates applied).
The problem has shown up for me in two difficult to track down ways:
1. I was running an AWS CLI rake task that was failing with access denied errors
2. Encountered errors running `sudo apt-get update` with errors saying that access to ubuntu repositories would be allowed in a few days.
Running `sudo hwclock --hctosys` or `sudo ntpdate pool.ntp.org` fixed it.

**Comment by TyIsI** [comment]
Ran into this today running `apt update` with WSL2 and Ubuntu 20.04.

Implemented the following work-around:

/etc/sudoers:
```
ALL ALL=NOPASSWD: /usr/local/bin/force_time.sh
```

/usr/local/bin/force_time.sh:
```
#!/bin/sh

if [ "$(whoami)" != "root" ] ; then
        echo "Not root, running with sudo..."
        sudo $0
else
        echo "Fixing clock from hardware clock"
        /usr/sbin/hwclock -s
fi
```

This will make it easy to implement in a `.bash_profile` or other startup script.

**Comment by djensen47** [comment]
@therealkenc 
> Something became much better in 18970 circa [August](https://github.com/microsoft/WSL/issues/4245#issuecomment-526363427) 2019, which means that fix shipped in 2004 aka 19041. That thing might not be your thing; feel free to revisit in a new issue with your repro.

I'm running 19041 and I'm still seeing this issue. I think this should be re-opened unless another ticket is tracking it. Thanks!

**Comment by onomatopellan** [comment]
@djensen47 This issue was already fixed but there is another similar problem in some scenarios. Thread and explanation in  https://github.com/microsoft/WSL/issues/5324#issuecomment-638348776

**Comment by djensen47** [comment]
It doesn't _seem_ fixed 🤷‍♂️ . The system time is out of sync in all scenarios for me: after a refresh, after waking from sleep. It just doesn't work. In this example WSL2 is ahead of Windows by 9 seconds. It's actually more because it took my a few seconds to create the screenshot/

![image](https://user-images.githubusercontent.com/527865/86746652-0e3a9d00-bff0-11ea-902c-f703c484bde8.png)

**Comment by onomatopellan** [comment]
Try with `wsl.exe --shutdown` and reopen the terminal. If you see the correct date then it's #5324.

**Comment by PaulaADiazSilva** [comment]
> Quick fix! In PowerShell or Command Prompt:
> 
> ```
> wsl --shutdown
> ```
> 
> Next time you start a WSL session, it'll be a cold boot and the time will be correct.

This works! thank you very much

**Comment by nicohouillon** [comment]
not fixed, still happening Microsoft Windows [Version 10.0.19041.388] 
returns errors with AWS api being out of sync

**Comment by pascoemitch** [comment]
Can confirm it is still happening on 19041.388
Everytime I wake after sleep or hibernation the clock is in the past.
Causes Github issues & AWS issues
Running `sudo ntpdate pool.ntp.org` solves it as a workaround.

**Comment by onenachi** [comment]
Can confirm it's still present on Windows 10 2004 (19041.388). Gnu Gmake gives time skew errors.

**Comment by Ezra** [comment]
Just got this myself on 19041.388 as well. Didn't get any of the ntpwhatever solutions working, but the `sudo hwclock -s` updated it just fine, so I guess I'll be working around with that until this is fixed.

**Comment by Ezra** [comment]
I guess we're supposed to be looking in #4114 and #5324 for this problem now?

**Comment by KyleRAnderson** [comment]
I still get this reliably every single time I wake the PC from hibernate. The `sudo hwclock -s` command fixes it, but really annoying.

I'm on a freshly baked Ubuntu distro (Ubuntu 20.04 right now) just after receiving the Windows 2004 update.

**Comment by anthony-steele-cko** [comment]
I have spent hours tracking missing metrics issues, only to eventually find that they're being logged _16 hours 13 minutes behind_ because the metrics service is running in a Linux container.  This issue has all kinds of knock-on effects.

**Comment by adamenger** [comment]
This is having various effects on all sorts of software that we're attempting to test on WSL. Running `sudo hwclock -s` fixed it but this is not a tenable approach long term.

**Comment by tommiv** [comment]
While this being fixed, one can use `chrony` as a workaround. It looks like you can't autostart it without systemd, but you can add `sudo chronyd` to your `.bashrc` / `.zshrc`. `chronyd` handles multiple run attempts automatically and returns really fast if it's already running (3ms on my machine) so it shouldn't make any visible impact on your terminal startup performance. I effectively forgot about any time sync issues after I set it up. The only drawback of this workaround – you have to modify your sudoers file to allow run `sudo chronyd` without password, but as for me it's not a big deal.

**Comment by onenachi** [comment]
Is this even being fixed? I see the issue is closed.

**Comment by garyo** [comment]
There's another issue on this topic at https://github.com/microsoft/WSL/issues/5324, still open since it is not fixed yet.

**Comment by min-xu-ai** [comment]
I am on the latest build (19041.450) and this doesn't seem to be fixed in ubuntu 20.04 with wsl2. I had to use `sudo hwclock -s` command as suggested above to workaround this.

**Comment by josefsabl** [comment]
Same issue, not fixed atm.

**Comment by padevel** [comment]
Cheers @tommiv 

> While this being fixed, one can use `chrony` as a workaround. It looks like you can't autostart it without systemd, but you can add `sudo chronyd` to your `.bashrc` / `.zshrc`. `chronyd` handles multiple run attempts automatically and returns really fast if it's already running (3ms on my machine) so it shouldn't make any visible impact on your terminal startup performance. I effectively forgot about any time sync issues after I set it up. The only drawback of this workaround – you have to modify your sudoers file to allow run `sudo chronyd` without password, but as for me it's not a big deal.

For those as new to this as I am, you might need to raise or disable the maximum skew limit. (Sleep durations will frequently exceed the default skew limit. From `/etc/chrony/chrony.conf` (Debian), comment out the `maxupdateskew` line:
```
# Stop bad estimates upsetting machine clock.
# maxupdateskew 100.0
```

**Comment by knuckolls** [comment]
The chrony workaround got me unblocked. In my case it was as simple as this on Ubunu 20.04 in WSL2

```
sudo apt-get update && sudo apt-get install -y chrony && sudo chronyd
```

**Comment by kellymilligan** [comment]
This is quite problematic when working with git, it seems that system time is used for the commit date. 

I can commit something to a GitHub repo, and it immedately shows it as being in the past. In my case, immediately as "committed 3 days ago", since I haven't restarted in 3 days. This makes for quite a confusing git log, and would require rewriting history on these commits to resolve 😬

**Comment by nate2014jatc** [comment]
Still present in W10 2004 (Build 19041.508).

**Comment by rupertparry** [comment]
Just adding to the chorus, as this hasn't been solved in a year. I've run into some really hard to diagnose errors using Google Cloud because of this.

**Comment by eriky** [comment]
This happened to me right after installing Docker Desktop with its WSL integration. I've seen others mention Docker as well. Perhaps there's a pattern here?

**Comment by danwhitston** [comment]
ARGH. Every time I commit and push changes, they show up as hours or days out of date. This is a complete mess.

**Comment by danwhitston** [comment]
I also can't use apt-get update as WSL 2 bash thinks it won't be valid until two and a half days in the future. Installing and running chronyd isn't helping. Installing and running ntpdate is throwing errors about connecting. Running the event setup at https://github.com/stuartleeks/wsl-clock also did nothing. I'm going to try rebooting the computer, see if that helps, because closing and restarting the terminal does nothing.

**Comment by eriky** [comment]
Please all, use the simple and quick fix as mentioned before as long as this issue is not fixed:

`sudo hwclock -s`

**Comment by tommiv** [comment]
@danwhitston it's enough to shutdown the VM with `wsl --shutdown`, no need to reboot. Also chronyd MUST help, as well as `sudo hwclock -s`. To make chrony more forgiving, take a look at this comment https://github.com/microsoft/WSL/issues/4245#issuecomment-687937163

**Comment by danwhitston** [comment]
Thanks both. You're right, chronyd must have been hitting the skew limit, given WSL's time was out by days rather than milliseconds.

**Comment by anthony-steele-cko** [comment]
That's funny, I had an incorrect date in WSL on waking my laptop today. 
So it does not seem to be "fixed" at all.

**Comment by kevinbosman** [comment]
`sudo hwclock -s` does not correct the time for me.

**Comment by YoraiLevi** [comment]
I have found this issues because this still happens. I was actually writing a topic about the `date` command and found this behavior by chance, hadn't tried the `hwclock -s`, but shutting down and restarting it works

**Comment by miklevin** [comment]
This is not fixed. I reset my clock every day.

**Comment by einarpersson** [comment]
This is not fixed. Just keep using the `sudo hwclock -s`

**Comment by grtrout** [comment]
I can't believe this is still broken. Debating about implementing some kind of hack to mitigate it, but in the meantime I just run `sudo hwclock -s` occasionally (particularly if I'm about to commit/push code). You would think something like this would be prioritized...surely the WSL maintainers themselves run into this issue all the time.

**Comment by Roritharr** [comment]
Just ran into this same problem. Why this issue is marked as closed is beyond me.

**Comment by garyo** [comment]
There's another issue on this topic at #5324, still open since it is not fixed yet.

**Comment by airglow923** [comment]
It does not seem to occur as of Windows 20H2 update. However, because it is not listed in the [resolved issues in Windows 10 version 20H2](https://docs.microsoft.com/en-us/windows/release-information/resolved-issues-windows-10-20h2), I am unsure of why it does not recur after the update.

**Comment by tomqwpl** [comment]
I also have this problem. The time reported by Windows appears accurate, seems to match "time.io" for example.
The time reported within WSL2 is about 25 seconds ahead.
The `sudo hwclock -s` command makes no difference for me, the time within WSL2 is unchanged. Restarting wsl with `wsl --shutdown` has made the clocks synchronize though.

**Comment by jasonatball** [comment]
Is there a solution if ```hwclock -s``` doesn't work:
```
root@6d5befe8cf47:/data# hwclock --verbose
hwclock from util-linux 2.33.1
System Time: 1607018356.723684
Trying to open: /dev/rtc0
Trying to open: /dev/rtc
Trying to open: /dev/misc/rtc
No usable clock interface found.
hwclock: Cannot access the Hardware Clock via any known method.
root@6d5befe8cf47:/data# cat /sys/devices/system/clocksource/clocksource0/available_clocksource
hyperv_clocksource_tsc_page
root@6d5befe8cf47:/data#
```

**Comment by tomqwpl** [comment]
Yes, I also get that error. So using ``hwclock -s`` on its own produces no output to say its failed, but ``--versose`` produces:

```
hwclock from util-linux 2.34
System Time: 1607018493.397590
Trying to open: /dev/rtc0
No usable clock interface found.
hwclock: Cannot access the Hardware Clock via any known method.
```

Of course I realise that you need to run it as ``sudo hwclock -s`` and that then does work (for me):

```
hwclock from util-linux 2.34
System Time: 1607018597.374906
Trying to open: /dev/rtc0
Using the rtc interface to the clock.
Assuming hardware clock is kept in UTC time.
Waiting for clock tick...
...got clock tick
Time read from Hardware Clock: 2020/12/03 18:03:18
Hw clock time : 2020/12/03 18:03:18 = 1607018598 seconds since 1969
Time since last adjustment is 1607018598 seconds
Calculated Hardware Clock drift is 0.000000 seconds
2020-12-03 18:03:16.947133+00:00
```

**Comment by jasonatball** [comment]
I was already root. I exec into a redis docker container: ```docker exec -it redis bash``` and at that point I'm root but I can't execute hwclock.

**Comment by ryanatball** [comment]
Open a bash shell directly from Windows, then try "sudo hwclock -s".   That runs but still left me a few seconds behind though.

**Comment by Mereep** [comment]
Same problem. It is December '20. The bug is from June '19.

This thing is serious since Docker is using WSL 2 now, effectively having wrong clocks in all containers.

**Comment by jurij** [comment]
prod was broken. solution was do delete windows and install unix system

**Comment by davystrong** [comment]
Just after spending a whole day trying to work out why a Docker container was getting 403 when connecting to AWS... then checked the date! Very annoying. Why has this issue been closed?

**Comment by arcs-** [comment]
Bump?

**Comment by tomqwpl** [comment]
Docker containers having the wrong clock (which is the effect of this issue) is a serious problem. The reason I ran into this was that I have a docker container running a service that is issuing authentication tokens for a web service. Trouble is that you authenticate and get a token, but when you use the token the receiver (not running in docker, and hence having a different clock) complains that the token is being used before it's issued because the clocks aren't in sync. This is just one example, but there would be many ways in which not having the clocks synched will cause unexpected issues.

**Comment by pauljfx** [comment]
Experiencing the same issue, this did not work for me
```
sudo hwclock -s
```

this works for some time but then the time resets again
```
sudo ntpdate pool.ntp.org
```

**Comment by RoyPour** [comment]
If you have a Docker container, don't try changing time within the container. Instead, change the server time with:
sudo ntpdate pool.ntp.org
(if you don't have it on system then install with: sudo apt-get update & sudo apt-get install ntpdate)
or 
sudo ~/set-date.sh
given set-date.sh is defined below.

* Having correct time becomes a must when you are pushing or pulling updates to Azure blob or any other data factory.


# ~/set-date.sh

# Get time from google server or some other reliable source
TIME_FROM_SERVER=$(curl -v --insecure --silent https://google.com/ 2>&1 | grep Date | sed -e 's/< Date: //');

# Set stored time
date +"%d%m%Y%H%M%S" -d "$TIME_FROM_SERVER";

echo "Time updated with success!";

**Comment by OscBacon** [comment]
Hey, I hope this can help :)
I decided to set up a crontab to update my time every 10 minutes.
I think that it's a good way to make sure the time doesn't drift too much.
Note that since you need sudo to run ntpdate,  you need to use the sudo crontab:

so, run `sudo crontab -e`, then add:
`10 * * * * ntpdate time.windows.com `

**Comment by sjschultze** [comment]
Folks still commenting on this issue might want to comment on #5324 instead because that one's still open and comments there are probably more likely to get noticed by MSFT than comments on this closed issue.

I can't believe this bug has existed for this long.  Wow.
